Localize

Deardrops opened the pull request:
TODOs:
- [ ] rebuild **CardInfo**

webxoss committed:
add hjson loader

Deardrops committed:
add map_xx.hjson

Deardrops committed:
flat hjson data struct

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  GUARD_FALSE: No
+  LIMITING: %s Only
+  WAIT: Waiting...
+  CLASS_SEPARATOR: ":\n"
```
一些特殊字符开头的文本必须用双引号括起来，如 :

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  CROSS_AND: >Cross< %s Right and %s Left
+  CROSS_OR:  or 
+  GUARD_DESCRIPTION: "[Guard] (By discarding this card from your hand, disable one of a LRIG's attacks)"
+  MULTI_ENER_DESCRIPTION: "[Constant]: [Multi Ener] (When you pay a cost, treat this card as if it has all colors)"
```
一些特殊字符开头的文本必须用双引号括起来，如 [

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  WAIT: Waiting...
+  CLASS_SEPARATOR: ":\n"
+  CROSS_NAME_QUOTE_LEFT: "\""
+  CROSS_NAME_QUOTE_RIGHT: "\""
```
表示包含一个"的字符串

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  精元: Origin
+}
+// TODO: remove DOM
+DOM: {
```
DOM将会在最后移除

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  FAILED_TO_PARSE_CODE: Failed to parse code!
+  CONFIRM_DELETE_DECK: Are you sure to delete "%s"?
+}
+DeckPage: {
```
暂定这种方式命名，类似的还有MainPage，BattlePage

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  CONFIRM_DELETE_DECK: Are you sure to delete "%s"?
+}
+DeckPage: {
+  new_deck: New Deck
```
比较倾向于_连接的变量名，全小写

Deardrops commented on src/lang/map_en.hjson:
```diff
@@ -1,360 +1,370 @@
+  delete: Delete
+  import: Import
+  export: Export
+  CONFIRM_DELETE: Are you sure to delete?
```
不大喜欢这样全大写的变量名，文本和单词可以用同样的**小写+_**形式命名

webxoss commented:
除了卡片信息的，全部删掉。留着干嘛

webxoss committed:
remove useless items

Deardrops committed:
Merge pull request #58 from webxoss/review

remove useless items

Deardrops committed:
improve localize

1.  rename map_xx,hsjon to xx.hsjon
zh_CN -> zh_Hans   zh_TW -> zh_Hant (not done)
2.  recovery '_misc' in hjson, needed by card info
3.  try use localize in DeckModal & Modal,  need further discussion

Deardrops commented on src/components/DeckModals.vue:
```diff
@@ -1,6 +1,8 @@
 import { mapGetters, mapState } from 'vuex'
 import Modal from 'components/Modal'
+import Localize from 'js/Localize'
+let $ = Localize.deckPage
```
简写函数名，type α

Deardrops commented on src/components/Modal.vue:
```diff
@@ -1,4 +1,8 @@
+import Localize from 'js/Localize'
+let $ = (text) => {
+  return Localize('common', text)
+}
```
简写函数名, type β (比较倾向于这种方式

Deardrops commented on src/components/Modal.vue:
```diff
@@ -18,10 +22,10 @@ export default {
     },
     cancleText() {
-      return this.config.cancleText || 'CANCLE'
+      return this.config.cancleText || $('CANCEL')
```
想把hjson中所有key改成小写的形式

webxoss replied:
嗯，大写太C了，小写好看

Deardrops commented on src/components/DeckModals.vue:
```diff
@@ -46,10 +48,10 @@ export default {
           defaultInput: this.deckName,
           validate: name => name && !this.deckNames.includes(name),
-          okText: 'RENAME',
+          okText: $('rename').toUpperCase(),
```
hjson中单词的value都是小写，需要大写的地方这样添加大写

Deardrops committed:
improve localize

rename zh_CN -> zh_Hans zh_TW -> zh_Hant
remove ALL useless function in localize.js
use L('ok') as abbreviation to Localize('common', 'ok')
use lower-case key-value in hjson (not done)
more further discussion needed

Deardrops committed:
improve localize

hjson: remove common, rename deckPage -> page

webxoss commented on src/components/AppHeader/HeaderMenu.vue:
```diff
@@ -89,6 +89,7 @@ export default {
 
   & > * {
     vertical-align: middle;
+    text-transform: capitalize;
```
放到外面 .item 那里吧？

Deardrops committed:
rebuild localize.js

use es6 feature
replace Localize('XXnamespace','XXlabel') by
Localize.XXnamespace('XXlabel')
Localize('XXlabel') === Localize.page('XXlabel')

webxoss commented on src/js/Localize.js:
```diff
@@ -1,86 +1,82 @@
-import { toArr, concat } from 'js/util.js'
 const Localize = (() => {
```
。。。模板化了，改

webxoss commented on src/js/Localize.js:
```diff
@@ -1,86 +1,82 @@
+  //
+  // usage: Localize.namespace('label')
+  //   e.g. Localize.color('red')
+  for (let namespace of Object.keys(Localize.map)) {
```
~~`for (let namespace of map)`~~

```js
for (let namespace in map)
```

webxoss commented on src/js/Localize.js:
```diff
@@ -1,86 +1,82 @@
+        return label
+      }
+      str = Localize.formatString(str, values)
+      str = Localize.traditionalize(str)
```
判断呢

Deardrops replied:
这里不需要判断吧，formatString和traditionalize都不会返回空字符的

webxoss commented on src/js/Localize.js:
```diff
@@ -1,86 +1,82 @@
-    for (let i = 0; i < str.length; i++) {
-      let char = str.charAt(i)
-      // let charCode = str.charCodeAt(i)
+    for (let char of str) {
```
不知道 babel 把这个转换成什么了，调查一些，糟糕就别用

Deardrops replied:
for..of 把可以识别`0xFFFF`编码以上的字符，运行起来没有问题。

webxoss replied:
[Try it out](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Ces2015-loose&targets=&browsers=&builtIns=false&experimental=false&loose=false&spec=false&code=for%20(let%20chr%20of%20'')%20%7B%7D)

大量长文本可能会卡

webxoss reviewed (commented):
没仔细看，自己测一下其他语言

Deardrops committed:
modularize Localize.js

Deardrops commented:
测试结果：英文，中文（繁体/简体）没有问题。其他语言由于缺少翻译文本无法显示（已全部设置为英文）

顺便，circleci在npm install的时候就失败了是什么鬼。。。

![image](https://cloud.githubusercontent.com/assets/9463871/23089335/298d8992-f5c2-11e6-981f-337ecaddc9fc.png)
如图所示

webxoss commented:
package.json 里 hjson 的版本改成 `^1.7.4` 吧

Deardrops committed:
improve Localize

1. extract function tranditional in Localize.js
2. store language in vuex.state
3. fix grammer in Rules.js
4. change hjson version to "^1.7.4"

webxoss commented on src/store.js:
```diff
@@ -150,4 +158,12 @@ const store = new Vuex.Store({
 })
 
+//watch language change && set Localize
+store.watch(state => {
```
`state`是 immutable 的，必须通过 `mutation` 改变，这里 watch 有意义吗？

webxoss replied:
好像也可以，太困了，脑子转不起来了，明天再细看

webxoss commented on src/js/traditionalize.js:
```diff
@@ -0,0 +1,20 @@
+
+// used by Localize only
+export default function traditionalize(str) {
+  if (!this.map.traditional) {
```
`this`?

Deardrops commented:
Close , ALL done by #64
